### PR TITLE
Fix perf hit from WinHttpGetProxyForUrl

### PR DESCRIPTION
**Pull request by caesar1995**
This PR optimizes some scenarios which we need to call `WinHttpGetProxyForUrl`.

1. Introduce the `WinHttpDetectAutoProxyConfigUrl` function, which implements a subset of the WPAD protocol: it attempts to auto-detect the URL for the proxy auto-config file, without downloading or executing the PAC file.

- If not able to find the script location. It is possible that there is no proxy auto-configuration file because the client has a direct Internet connection, and does not need a proxy server. Or a proxy server may be required, but the local network may not support WPAD. In both cases, the proxy configuration has to be obtained from the user or found somewhere on the client machine. We will use the information from calling into `WinHttpGetIEProxyConfigForCurrentUser` to see whether a proxy server is listed in Internet Explorer's settings. Cache the information.

- If able to find the script location, the AutoDetect setting has precedence over the other settings. So if there is already a automatic configuration script location (from `WinHttpGetIEProxyConfigForCurrentUser`), this value will be overridden. We will cache the related information and provide it to `WinHttpGetProxyForUrl`.

2. If there is no proxy, the information has been cached from the above procedures, we will stop calling into `WinHttpGetProxyForUrl`.

We have exhaustive test cases verifying behaviors with different proxy settings. Please let me know if there are special cases we care about, and I will add tests for them.

https://github.com/dotnet/corefx/blob/85062446c4e5af0c205f646e0ae720e160b20e85/src/System.Net.Http.WinHttpHandler/tests/UnitTests/WinHttpHandlerTest.cs#L641-L832

Fix: #28543

### CODE REVIEW

**Comment by caesar1995** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

File: src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs [hunk 1]
```diff
@@ -310,5 +310,13 @@ public struct WINHTTP_WEB_SOCKET_STATUS
             public uint dwBytesTransferred;
             public WINHTTP_WEB_SOCKET_BUFFER_TYPE eBufferType;
         }
+
+        [Flags]
+        internal enum AutoDetectType
```
Comment (davidsh) [inline src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs]: nit: 'internal' should be 'public' like all the other nested classes/types in this class.  The enclosing class is already 'internal'

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 2]
```diff
@@ -178,5 +153,97 @@ public bool ManualSettingsOnly
...
+            AutoConfigUrl = SafeDetectAutoProxyUrl(Interop.WinHttp.AutoDetectType.Dhcp);
+
+            if (AutoConfigUrl == null)
+            {
+                WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Attempting discovery AUTO_DETECT_TYPE_DNS_A.");
+                AutoConfigUrl = SafeDetectAutoProxyUrl(Interop.WinHttp.AutoDetectType.DnsA);
+            }
+            else
+            {
+                // Auto-detect succeed.
```
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: 'succeed' -> 'succeeded'

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 3]
```diff
@@ -178,5 +153,97 @@ public bool ManualSettingsOnly
...
+
+            if (AutoConfigUrl == null)
+            {
+                WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Auto discovery failed.");
+                AutoDetect = false;
+                _scriptDetectionFailed = true;
+            }
+        }
+
+        private string SafeDetectAutoProxyUrl(Interop.WinHttp.AutoDetectType discoveryMethod)
```
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Why use 'Safe' as the prefix of this method name?  Maybe call this method something like "GetAutoProxyUrl".

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs [hunk 4]
```diff
@@ -611,6 +611,24 @@ private static bool CopyToBufferOrFailIfInsufficientBufferLength(string value, I
             return true;
         }
 
+        public static bool WinHttpDetectAutoProxyConfigUrl(
```
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs]: Since you're adding this method (so that the unit tests keep compiling), it might be useful to also add some new unit tests to test your new code in this PR.

File: src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs (hunk 1 above)
Comment (caesar1995) [reply src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs]: Will change.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 2 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Will change.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 3 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Will do.

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs (hunk 4 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs]: Many existing test cases will hit the new code path. (They are exhaustive and cover almost every scenario)

For example, these test cases will hit `WinHttpDetectAutoProxyConfigUrl` returns `true`:

`SendAsync_NoAutomaticProxySupportAndUseWinInetSettingsWithAutoDetectSetting_ExpectedWinHttpProxySettings`
`SendAsync_NoAutomaticProxySupportAndUseWinInetSettingsWithAutoDetectButPACFileNotDetectedOnNetwork_ExpectedWinHttpProxySettings`
`SendAsync_NoAutomaticProxySupportAndUseWinInetSettingsWithAutoDetectSettingAndManualSettingButPACFileNotFoundOnNetwork_ExpectedWinHttpProxySettings`

These test cases will hit `WinHttpDetectAutoProxyConfigUrl` returns `false`:

`SendAsync_NoAutomaticProxySupportAndUseWinInetSettingsWithMissingRegistrySettings_ExpectedWinHttpProxySettings`
`SendAsync_NoAutomaticProxySupportAndUseWinInetSettingsWithEmptySettings_ExpectedWinHttpProxySettings`
`SendAsync_NoAutomaticProxySupportAndUseWinInetSettingsWithManualSettingsOnly_ExpectedWinHttpProxySettings`

The name `NoAutomaticProxySupport` is somewhat confusing, actually it means: when we pass in access type `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY` to `WinHttpOpen`, the function will return a invalid session handle. (For example, when running on a platform earlier than Win8.1/Win2K12R2). Then we will need to use WPAD related functions to retrieve proxy data.

I think the above 6 test cases covered all proxy setting scenarios we may want to call `WinHttpDetectAutoProxyConfigUrl`.

**Comment by caesar1995** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 5]
```diff
@@ -178,5 +153,97 @@ public bool ManualSettingsOnly
...
+                return;
+            }
+
+            WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Attempting discovery PROXY_AUTO_DETECT_TYPE_DHCP.");
+            AutoConfigUrl = GetAutoProxyConfigUrl(Interop.WinHttp.AutoDetectType.Dhcp);
+
+            if (AutoConfigUrl == null)
+            {
+                WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Attempting discovery AUTO_DETECT_TYPE_DNS_A.");
+                AutoConfigUrl = GetAutoProxyConfigUrl(Interop.WinHttp.AutoDetectType.DnsA);
```
Comment (davidsh) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: You don't need to call this twice.  The AutoDetectType enum values are flags.  So, just OR them together and use a single call to WinHttpDetectAutoProxyConfigUrl.

File: src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs [hunk 6]
```diff
@@ -310,5 +310,13 @@ public struct WINHTTP_WEB_SOCKET_STATUS
             public uint dwBytesTransferred;
             public WINHTTP_WEB_SOCKET_BUFFER_TYPE eBufferType;
         }
+
+        [Flags]
```
Comment (davidsh) [inline src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs]: For flag bits like this in p/invoke's, we usually don't make them enum type.  We simply make them `public const uinit` and keep the native names.  

See for example: https://github.com/dotnet/corefx/blob/master/src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs#L108

So, this would probably be done as this:

```c#
public const uint WINHTTP_AUTO_DETECT_TYPE_DHCP  = 0x00000001;
public const uint WINHTTP_AUTO_DETECT_TYPE_DNS_A = 0x00000010;
```

And then you would reference it in code with `Interop.WinHttp.WINHTTP_AUTO_DETECT_TYPE_DHCP`, etc.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 7]
```diff
@@ -178,5 +153,97 @@ public bool ManualSettingsOnly
 
             return useProxy;
         }
+
+        private void DetectScriptLocation()
+        {
+            if (_scriptDetectionFailed || AutoConfigUrl != null)
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: DetectScriptLocation is only called from the ctor, at the very beginning.  How could _scriptDetectionFailed be true?  You can make this a Debug.Assert(!_scriptDetectionFailed) if you want to validate that, but I don't think that's even necessary.

Comment (stephentoub) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Same comment for AutoConfigUrl... how could it be non-null at this point?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 8]
```diff
@@ -178,5 +153,97 @@ public bool ManualSettingsOnly
...
+            {
+                // Auto-detect succeeded.
+                return;
+            }
+
+            if (AutoConfigUrl == null)
+            {
+                WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Auto discovery failed.");
+                AutoDetect = false;
+                _scriptDetectionFailed = true;
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Why do we need _scriptDetectionFailed at all?  It seems like it's true iff AutoConfigUrl.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 9]
```diff
@@ -14,44 +14,18 @@ namespace System.Net.Http
...
-                WinHttpTraceHelper.Trace("WinInetProxyHelper.ctor: _useProxy={0}", _useProxy);
-            }
-
-            finally
-            {
-                // FreeHGlobal already checks for null pointer before freeing the memory.
-                Marshal.FreeHGlobal(proxyConfig.AutoConfigUrl);
-                Marshal.FreeHGlobal(proxyConfig.Proxy);
-                Marshal.FreeHGlobal(proxyConfig.ProxyBypass);
+                GetIEProxySetting();
```
Comment (stephentoub) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: So the WinHttp-detected script location takes precedence over the user's local configuration? Is that correct? Or does WinHttp factor that in, too?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 8 above)
Comment (stephentoub) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: And do we need AutoDetect? In what situation will that be true but not have a url, and what are we supposed to do at that point?

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs (hunk 4 above)
Comment (stephentoub) [reply src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs]: > I think the above 6 test cases covered all proxy setting scenarios we may want to call 

How are you validating the script cases? Are you doing that manually locally by setting up some working auto-config location?

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 9 above)
Comment (davidsh) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: >So the WinHttp-detected script location takes precedence over the user's local configuration? 

AutoDetect has precedence over the other settings.  If you look at the below picture, you see that there can be multiple settings for determining where the PAC file (if any) and then if there is a specific proxy server present.  Each checkbox represents what to use to find/use PAC file or specfic proxy in order of priority.  If a step fails, the fallback is usually to use the next setting (checkbox).

image

So, "Automatically detect settings" is what calling WinHttpDetectAutoProxyConfigUrl() helps figure out (we should only be calling that if the checkbox is true and we only know that by calling 
WinHttpGetIEProxyConfigForCurrentUser() in the first place.

I will say, though, that AutoDetect could fail (i.e WPAD protocol fails) but that doesn't mean there is no PAC file, nor that we should stop calling WinHttpGetProxyForUrl() for each request.

File: src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs (hunk 4 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/tests/UnitTests/FakeInterop.cs]: > How are you validating the script cases?

There are three cases related to script: 1. discover the PAC URL, 2. download the PAC script file, 3. execute the script and return the proxy configuration. `WinHttpDetectAutoProxyConfigUrl` is only able to achieve case 1: to auto-detect the URL for the proxy auto-config file (without downloading/executing), and we are reply on `WinHttpGetProxyForUrl` to finish case 2 and 3.

So I assume you are asking for the first case, discover the PAC URL. There are two ways to achieve it: 

1. Automatically detect settings (using WPAD protocol, which I think it's hard to setup testing, but @davidsh can confirm?) We can trust `WinHttpDetectAutoProxyConfigUrl` to return the correct PAC URL. (if it exists)
2. Use automatic configuration script. I have tested this scenario and set up a working proxy auto-config location (on a IIS server), the `WinHttpGetIEProxyConfigForCurrentUser` will return the correct URL in this case, and store its value in `AutoConfigUrl`.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 5 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: I copied the pattern from Framework, and I think this approach is slightly better than OR them together: assume `WinHttpDetectAutoProxyConfigUrl` succeed, by separating the operations, we will know exactly which protocol is used to locate the PAC file, and log the information.

File: src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs (hunk 6 above)
Comment (caesar1995) [reply src/Common/src/Interop/Windows/winhttp/Interop.winhttp_types.cs]: Will change to follow the same pattern.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 8 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: > And do we need AutoDetect? In what situation will that be true but not have a url, and what are we supposed to do at that point?

It's common that `AutoDetect` is true while `AutoConfigUrl` is null. This can be caused by, specifying `Automatically detect settings` while not providing a location for `automatic configuration script` in the IE configuration.

If `AutoConfigUrl` is null (no configuration script is specified), and `AutoDetect` is true, the next step is that we will use WPAD to find the script location. (`WinHttpGetProxyForUrl`/`WinHttpDetectAutoProxyConfigUrl`).

Since within current function, we are calling into `WinHttpDetectAutoProxyConfigUrl` to help us find the script. If it fails, we will update the `AutoDetect` to false, so that we will not calling into `WinHttpGetProxyForUrl` anymore. (The purpose of this PR, fix the perf hit for `WinHttpGetProxyForUrl`).

(we will call into `WinHttpGetProxyForUrl` if `if (_proxyHelper != null && _proxyHelper.AutoSettingsUsed)` -> `AutoSettingsUsed: return AutoDetect || !string.IsNullOrEmpty(AutoConfigUrl);`)

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 9 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: > So the WinHttp-detected script location takes precedence over the user's local configuration?

This is not the case for Windows 8.1 and newer. We will only using WinHttp-detected script location if failed to get the user's local configuration. Here is the proof:

1. We do a [`WinHttp.WinHttpOpen` with parameter `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY`](https://github.com/dotnet/corefx/blob/master/src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpHandler.cs#L716) (if no proxy specified & we need to detect it ourselves) before we create the `WinInetProxyHelper` object.
2. When we specify `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY` flag for `WinHttp.WinHttpOpen`, it will use system and per-user proxy settings (including the Internet Explorer proxy configuration) to determine which proxy/proxies to use. (https://msdn.microsoft.com/en-us/library/windows/desktop/aa384098(v=vs.85).aspx)
3. If the above procedure failed, then we will try to create the `WinInetProxyHelper`, and do the winhttp detecting script location stuff.

The `WinHttpGetIEProxyConfigForCurrentUser` functionality duplicated with `WinHttpOpen`. However, notice that the flag `WINHTTP_ACCESS_TYPE_AUTOMATIC_PROXY` is only supported in Windows 8.1 and newer. Therefore, we still need to calling into `WinHttpGetIEProxyConfigForCurrentUser` for previous Windows version.

I do agree that on OS before Windows 8.1, we may have issues with the ordering (as David has pointed out: we only know if `Automatically detect settings` is on by calling WinHttpGetIEProxyConfigForCurrentUser() in the first place). I will change the code to correct the behavior.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 7 above)
Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Good catch, while I was writing this function in the first place, the design was that it could be call multiple times. Adding cache like this will save time calling into it.

Later after refactoring the code and moving it to the ctor, it only serves the purpose that to help us determine if a proxy is used or not. The logic here is not useful at all. Will remove it.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs (hunk 5 above)
Comment (davidsh) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: It is most likely going to be slower to make separate calls.  And the only benefit is being able to log more details.  I don't think that is a sufficient reason to make the product code slower all the time.

Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: Will change.

File: src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs [hunk 10]
```diff
@@ -178,5 +151,85 @@ public bool ManualSettingsOnly
...
+                    {
+                        WinHttpTraceHelper.Trace("WinInetProxyHelper.DetectScriptLocation: Auto discovery failed.");
+
+                        // This is to improve performance. For example, on home networks, where auto-detect will always
+                        // fail, but IE settings turn auto-detect ON by default. So in home networks on each
+                        // call we would try to retrieve the PAC location.
+
+                        // The downside of this approach is that if after some time the PAC file can be downloaded,
+                        // we will skip it.
+                        AutoDetect = false;
```
Comment (caesar1995) [inline src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: As the email discussion with WinHttp team suggest, "The answer is if you want to be correct, you have to call the API (`WinHttpGetProxyForUrl`). Without re-implementing half of the proxy code in WinHttp and calling into tons of other components, you cannot know when it is Ok to skip it."

We will reach here when: 1. `WinHttpOpen` returns a invalid session handle, 2. `WinHttpGetIEProxyConfigForCurrentUser` not able to get automatic configuration script, 3. `WinHttpDetectAutoProxyConfigUrl` unable to auto discover the PAC URL (at this moment). It's highly likely that, there is no proxy used. But we are not 100% sure.

As the comment has suggested, there is a potential risk to set `AutoDetect = false` (which can cause `AutoSettingsUsed` returns false and skip calling into `WinHttpGetProxyForUrl`). On framework we decided to take the risk, and it already caused some issues there.

I'd like to hear people's feedback on the optimization in this case.

Comment (davidsh) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: >I'd like to hear people's feedback on the optimization in this case.

Can we quantify what the performance hit is if we always call the `WinHttpGetProxyForUrl` API for each request. It's hard to make tradeoff decisions if we don't understand what we're comparing exactly.

WinHttpGetProxyForUrl() already does internal caching of the PAC file whether it was found using WPAD protocol or explicitly specified. So, we aren't repeatedly doing WPAD protocol to find a PAC file for every request. So, I don't understand what the performance hit is in the steady-state case by calling this API. 

In cases where WinHttpGetIEProxyConfigForCurrentUser() has returned settings where "Automatically detect settings" or "Use automatic configuration script" are unchecked, it means that WPAD protcol shouldn't be used, and there is no explicit PAC file set.  So, that seems like the only case where we can avoid calling `WinHttpGetProxyForUrl` for each request.  The only other useful thing about WinHttpGetProxyForUrl would be to parse the proxy bypass list as specified in IE settings.  But, we decided to manually parse that list using managed code now.

We only call WinHttpGetIEProxyConfigForCurrentUser() once when the handler is created. That means that if the IE settings change while an application is still running, we won't be able to react to that.  In that case, only HttpClient/handler recycling can solve that.

Bottom line, I think trying to use WinHttpDetectAutoProxyConfigUrl() as a way to avoid calling WinHttpGetProxyForUrl() is not the safest approach in this PR.  I would prefer to use the algorithm I mentioned above regarding IE settings to determine whether or not we can skip calling WinHttpGetProxyForUrl().

Comment (caesar1995) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: > I think trying to use WinHttpDetectAutoProxyConfigUrl() as a way to avoid calling WinHttpGetProxyForUrl() is not the safest approach in this PR.

I agree. I don't feel comfortable doing this as well, mainly because we don't know for sure when is ok to skip the `WinHttpGetProxyForUrl` call, which may result in some hard bugs in the future.

I will take the safe way and depend on IE settings. Will close this PR and submit a new one.

Comment (davidsh) [reply src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinInetProxyHelper.cs]: >We need to make it as pay-for-play as we can.

I think if we want it pay-for-play, then the new PR #28965 covers that.  If a server wants optimum performance and avoid calling WinHttpGetProxyForUrl(), then it will make sure that the IE settings registry keys (if present) for "Use automatic configuration script" and "Automatically detect settings" are unchecked.
